File: public/js/midp/MIDlet.js
/** javax.microedition.midlet.MIDlet */
export class MIDlet {
  constructor() {
    this._props = Object.create(null);
    this._destroyed = false;
  }
  getAppProperty(key) {
    return this._props[key] ?? null;
  }
  notifyDestroyed() {
    this._destroyed = true;
    console.info('[MIDlet] destroyed');
  }
  platformRequest(url) {
    if (url) window.open(url, '_blank');
  }
  // subclass hooks
  startApp() {}
  pauseApp() {}
  destroyApp(_unconditional) {}
}

File: teavm-port/src/main/java/javax/microedition/midlet/MIDlet.java

package javax.microedition.midlet;

public abstract class MIDlet {
    private boolean destroyed;

    protected MIDlet() {}

    protected abstract void startApp() throws MIDletStateChangeException;
    protected abstract void pauseApp();
    protected abstract void destroyApp(boolean unconditional) throws MIDletStateChangeException;

    public final void notifyDestroyed() { destroyed = true; }
    public final void notifyPaused() {}
    public final String getAppProperty(String key) { return null; }
    public final boolean platformRequest(String url) { return false; }
    public final int checkPermission(String permission) { return 0; }

    /** Web adapter entry */
    public final void startAppPublic() {
        try { startApp(); } catch (MIDletStateChangeException e) { throw new RuntimeException(e); }
    }
}
